File: backend-nodejs/aiAgents.js
const axios = require('axios');

const MISTRAL_API_KEY = process.env.MISTRAL_API_KEY;
const MISTRAL_API_URL = 'https://api.mistral.ai/v1/chat/completions';

// Call Mistral AI (text only)
async function callMistralAI(systemPrompt, userPrompt) {
  try {
    const response = await axios.post(
      MISTRAL_API_URL,
      {
        model: 'mistral-large-latest',
        messages: [
          { role: 'system', content: systemPrompt },
          { role: 'user', content: userPrompt }
        ],
        temperature: 0.7,
        max_tokens: 2000
      },
      {
        headers: {
          'Authorization': `Bearer ${MISTRAL_API_KEY}`,
          'Content-Type': 'application/json'
        },
        timeout: 30000
      }
    );
    return response.data.choices[0].message.content;
  } catch (error) {
    console.error('Mistral AI Error:', error.response?.data || error.message);
    return null;
  }
}

// Call Mistral Vision API for OCR
async function callMistralVision(base64Image, mimeType, textPrompt) {
  try {
    const imageUrl = `data:${mimeType};base64,${base64Image}`;
    const response = await axios.post(
      MISTRAL_API_URL,
      {
        model: 'pixtral-large-latest',
        messages: [
          {
            role: 'user',
            content: [
              { type: 'text', text: textPrompt },
              { type: 'image_url', image_url: { url: imageUrl } }
            ]
          }
        ],
        temperature: 0.1,
        max_tokens: 2000
      },
      {
        headers: {
          'Authorization': `Bearer ${MISTRAL_API_KEY}`,
          'Content-Type': 'application/json'
        },
        timeout: 60000
      }
    );
    return response.data.choices[0].message.content;
  } catch (error) {
    console.error('Mistral Vision Error:', error.response?.data || error.message);
    return null;
  }
}

// Extract salary information from uploaded document using Mistral Vision
async function extractSalaryInfo(base64Image, mimeType) {
  const prompt = `Analyze this salary slip / income document image carefully.
Extract the following information and return ONLY valid JSON:
{
  "employee_name": "name found or null",
  "gross_salary": number or null,
  "net_salary": number or null,
  "month": "month/period or null",
  "employer": "company name or null",
  "deductions": number or null,
  "verified": true if this looks like a legitimate salary document, false otherwise
}
If this is not a salary/income document, set verified to false.
Return ONLY the JSON, no other text.`;

  const result = await callMistralVision(base64Image, mimeType, prompt);

  if (result) {
    try {
      const jsonMatch = result.match(/\{[\s\S]*\}/);
      if (jsonMatch) {
        const parsed = JSON.parse(jsonMatch[0]);
        return {
          ...parsed,
          raw_response: result,
          extraction_method: 'mistral_vision'
        };
      }
    } catch (e) {
      console.error('Failed to parse salary extraction JSON:', e);
    }
  }

  // Fallback - document uploaded but couldn't be read
  return {
    employee_name: null,
    gross_salary: null,
    net_salary: null,
    verified: false,
    extraction_method: 'fallback',
    error: 'Could not extract salary information from document'
  };
}

// Verify face image using Mistral Vision
async function verifyFaceImage(base64Image, mimeType) {
  const prompt = `Analyze this image. Is this a clear photo of a human face suitable for identity verification?
Return ONLY valid JSON:
{
  "is_face": true or false,
  "clarity": "good", "fair", or "poor",
  "notes": "brief description"
}
Return ONLY the JSON.`;

  const result = await callMistralVision(base64Image, mimeType, prompt);

  if (result) {
    try {
      const jsonMatch = result.match(/\{[\s\S]*\}/);
      if (jsonMatch) {
        return JSON.parse(jsonMatch[0]);
      }
    } catch (e) {
      console.error('Failed to parse face verification JSON:', e);
    }
  }

  // Fallback - accept the upload
  return { is_face: true, clarity: 'fair', notes: 'Verification via fallback' };
}

// Credit Scoring Agent
async function creditScoringAgent(userData, loanData) {
  const systemPrompt = `You are a credit scoring expert. Analyze financial data and provide credit scores (300-850). 
Return ONLY valid JSON format with keys: credit_score (number), risk_level (low/medium/high), factors (array of strings).`;

  const userPrompt = `Analyze the following data and calculate credit score:

User Financial Data:
- Monthly Income: ₹${loanData.monthly_income?.toLocaleString('en-IN') || 'N/A'}
- Employment Type: ${loanData.employment_type || 'N/A'}
- Existing Loans: ${loanData.existing_loans || 0}
- Account Balance: ₹${userData.account_balance?.toLocaleString('en-IN') || 'N/A'}
- Past Loan History: ${userData.past_loans_count || 0} loans
- Total Past Transactions: ${userData.total_transactions || 0}

Loan Request:
- Amount: ₹${loanData.amount?.toLocaleString('en-IN') || 'N/A'}
- Purpose: ${loanData.purpose || 'N/A'}
- Tenure: ${loanData.tenure_months || 'N/A'} months

Return JSON with: credit_score, risk_level, factors.`;

  const result = await callMistralAI(systemPrompt, userPrompt);

  if (result) {
    try {
      const jsonMatch = result.match(/\{[\s\S]*\}/);
      if (jsonMatch) {
        return JSON.parse(jsonMatch[0]);
      }
    } catch (e) {
      console.error('Failed to parse credit score JSON:', e);
    }
  }

  // Fallback calculation
  let baseScore = 650;
  if (loanData.monthly_income > 50000) baseScore += 50;
  if (loanData.existing_loans === 0) baseScore += 40;
  if (userData.past_loans_count > 0) baseScore += 30;
  if (userData.account_balance > 100000) baseScore += 30;

  const risk = baseScore >= 720 ? 'low' : baseScore >= 650 ? 'medium' : 'high';

  return {
    credit_score: Math.min(850, baseScore),
    risk_level: risk,
    factors: ['Stable income', 'Low existing debt', 'Good transaction history']
  };
}

// Document Verification Agent
async function documentVerificationAgent(userData, faceVerification, salaryData) {
  let score = 0;
  const notes = [];

  // Aadhar check
  if (userData.aadhar) { score += 25; notes.push('Aadhar verified'); }
  // PAN check
  if (userData.pan) { score += 25; notes.push('PAN verified'); }
  // Face check
  if (faceVerification?.is_face) {
    score += 25;
    notes.push(`Face verified (clarity: ${faceVerification.clarity || 'N/A'})`);
  }
  // Salary check
  if (salaryData?.verified) {
    score += 25;
    notes.push('Salary slip verified via AI OCR');
  } else if (salaryData) {
    score += 15;
    notes.push('Salary slip uploaded but unverified');
  }

  return {
    aadhar_verified: !!userData.aadhar,
    pan_verified: !!userData.pan,
    face_verified: !!faceVerification?.is_face,
    salary_slip_verified: !!salaryData?.verified,
    verification_score: score,
    notes: notes.join('. ')
  };
}

// EMI Calculation Agent
function emiCalculationAgent(principal, tenureMonths, riskLevel) {
  const interestRates = { low: 8.5, medium: 11.0, high: 14.5 };
  const annualRate = interestRates[riskLevel] || 11.0;
  const monthlyRate = annualRate / 12 / 100;

  let emi;
  if (monthlyRate > 0) {
    emi = principal * monthlyRate * Math.pow(1 + monthlyRate, tenureMonths) /
          (Math.pow(1 + monthlyRate, tenureMonths) - 1);
  } else {
    emi = principal / tenureMonths;
  }

  const totalAmount = emi * tenureMonths;
  const totalInterest = totalAmount - principal;

  return {
    emi_amount: Math.round(emi * 100) / 100,
    interest_rate: annualRate,
    total_amount: Math.round(totalAmount * 100) / 100,
    total_interest: Math.round(totalInterest * 100) / 100,
    processing_fee: Math.round(principal * 0.01 * 100) / 100
  };
}

// Underwriting Agent
async function underwritingAgent(loanData, creditAnalysis, docVerification, emiCalculation) {
  const systemPrompt = `You are a senior loan underwriting officer. Review all analysis and make final loan approval decisions. 
Return ONLY valid JSON with keys: decision (approved/rejected), approved_amount (number), reason (string), conditions (array of strings).`;

  const userPrompt = `Loan Underwriting Review:

Loan Request:
- Amount: ₹${loanData.amount?.toLocaleString('en-IN')}
- Purpose: ${loanData.purpose}
- Tenure: ${loanData.tenure_months} months

Credit Analysis:
- Credit Score: ${creditAnalysis.credit_score}
- Risk Level: ${creditAnalysis.risk_level}
- Factors: ${creditAnalysis.factors?.join(', ')}

Document Verification:
- Verification Score: ${docVerification.verification_score}%
- Notes: ${docVerification.notes}

EMI Calculation:
- EMI Amount: ₹${emiCalculation.emi_amount?.toLocaleString('en-IN')}
- Interest Rate: ${emiCalculation.interest_rate}%
- Total Repayment: ₹${emiCalculation.total_amount?.toLocaleString('en-IN')}

Make underwriting decision. Return JSON with: decision, approved_amount, reason, conditions.`;

  const result = await callMistralAI(systemPrompt, userPrompt);

  if (result) {
    try {
      const jsonMatch = result.match(/\{[\s\S]*\}/);
      if (jsonMatch) {
        return JSON.parse(jsonMatch[0]);
      }
    } catch (e) {
      console.error('Failed to parse underwriting JSON:', e);
    }
  }

  // Fallback decision
  const creditScore = creditAnalysis.credit_score;
  const requestedAmount = loanData.amount;

  if (creditScore >= 700) {
    return {
      decision: 'approved',
      approved_amount: requestedAmount,
      reason: 'Strong credit profile and stable income',
      conditions: ['Maintain account balance', 'Timely EMI payments']
    };
  } else if (creditScore >= 650) {
    return {
      decision: 'approved',
      approved_amount: Math.round(requestedAmount * 0.85),
      reason: 'Moderate risk profile, approved with reduced amount',
      conditions: ['Provide guarantor', 'Monthly income proof', 'Timely EMI payments']
    };
  } else {
    return {
      decision: 'approved',
      approved_amount: Math.round(requestedAmount * 0.7),
      reason: 'Higher risk but manageable with conditions',
      conditions: ['Provide guarantor', 'Life insurance coverage', 'Quarterly income verification']
    };
  }
}

// Full Loan Processing Orchestrator (called after all documents collected)
async function processLoanApplication(userData, loanData, faceVerification, salaryData) {
  console.log('Starting multi-agent loan processing...');

  // Step 1: Credit Scoring
  const creditAnalysis = await creditScoringAgent(userData, loanData);
  console.log('Credit score:', creditAnalysis.credit_score);

  // Step 2: Document Verification
  const docVerification = await documentVerificationAgent(userData, faceVerification, salaryData);
  console.log('Document verification score:', docVerification.verification_score);

  // Step 3: EMI Calculation
  const emiCalculation = emiCalculationAgent(
    loanData.amount,
    loanData.tenure_months,
    creditAnalysis.risk_level
  );
  console.log('EMI calculated:', emiCalculation.emi_amount);

  // Step 4: Underwriting Decision
  const underwritingDecision = await underwritingAgent(
    loanData, creditAnalysis, docVerification, emiCalculation
  );
  console.log('Underwriting decision:', underwritingDecision.decision);

  return {
    credit_analysis: creditAnalysis,
    document_verification: docVerification,
    emi_calculation: emiCalculation,
    underwriting_decision: underwritingDecision,
    processed_at: new Date().toISOString()
  };
}

// Chatbot Agent - context-aware
async function chatbotAgent(userMessage, userContext, loanContext) {
  let loanInfo = '';
  if (loanContext) {
    loanInfo = `\n\nActive Loan Application:
- Loan ID: ${loanContext.loan_id}
- Amount: ₹${loanContext.amount?.toLocaleString('en-IN')}
- Purpose: ${loanContext.purpose}
- Status: ${loanContext.status}
- Face Verified: ${loanContext.face_verified ? 'Yes' : 'No'}
- Salary Slip Verified: ${loanContext.salary_verified ? 'Yes' : 'No'}`;
  }

  const systemPrompt = `You are a professional virtual loan officer at NBFC Bank. Your role:
- Help users understand loan products and eligibility
- Guide them through the application process
- Explain credit scores, interest rates, and EMI calculations
- Answer questions about existing loans and repayment
- Be professional, empathetic, and helpful

${loanContext ? `The user has an active loan application. Guide them through document collection:
- If face is NOT verified, ask them to upload their face photo using the camera button below the chat.
- If face IS verified but salary slip is NOT verified, ask them to upload their salary slip using the upload button.
- If both are verified, the loan is being processed by our AI agents.
- Do NOT ask the user to leave the chat to upload documents - they can upload directly here.` : ''}

Keep responses under 150 words. Be concise and action-oriented.`;

  const userPrompt = `User Context:
- Name: ${userContext.name}
- Account Balance: ₹${userContext.account_balance?.toLocaleString('en-IN')}
- Credit Score: ${userContext.credit_score || 'Not calculated yet'}
- Active Loans: ${userContext.active_loans || 0}
- Past Loans: ${userContext.past_loans_count || 0}${loanInfo}

User Message: ${userMessage}`;

  const result = await callMistralAI(systemPrompt, userPrompt);
  return result || "I'm here to help you with your loan queries. How can I assist you today?";
}

module.exports = {
  processLoanApplication,
  chatbotAgent,
  creditScoringAgent,
  emiCalculationAgent,
  extractSalaryInfo,
  verifyFaceImage
};

File: backend/ai_agents.py
import os
import json
from typing import Dict, Any, Optional
from emergentintegrations.llm.chat import LlmChat, UserMessage
from datetime import datetime, timedelta
import random

MISTRAL_API_KEY = os.environ.get("MISTRAL_API_KEY", "")

class CreditScoringAgent:
    """Analyzes user financial data and computes credit score"""
    
    def __init__(self):
        self.chat = LlmChat(
            api_key=MISTRAL_API_KEY,
            session_id=f"credit_scoring_{datetime.utcnow().timestamp()}",
            system_message="You are a credit scoring expert. Analyze financial data and provide credit scores (300-850). Return only JSON format with keys: credit_score, risk_level (low/medium/high), factors."
        ).with_model("openai", "gpt-5.2")
    
    async def calculate_score(self, user_data: Dict[str, Any], loan_data: Dict[str, Any]) -> Dict[str, Any]:
        prompt = f"""
Analyze the following data and calculate credit score:

User Financial Data:
- Monthly Income: ₹{loan_data.get('monthly_income', 0):,.2f}
- Employment Type: {loan_data.get('employment_type', 'Unknown')}
- Existing Loans: {loan_data.get('existing_loans', 0)}
- Account Balance: ₹{user_data.get('account_balance', 0):,.2f}
- Past Loan History: {len(user_data.get('past_loans', []))} loans
- Total Past Transactions: {user_data.get('total_transactions', 0)}

Loan Request:
- Amount: ₹{loan_data.get('amount', 0):,.2f}
- Purpose: {loan_data.get('purpose', 'Unknown')}
- Tenure: {loan_data.get('tenure_months', 0)} months

Return JSON with: credit_score, risk_level, factors (list of positive/negative factors).
"""
        try:
            response = await self.chat.send_message(UserMessage(text=prompt))
            result = json.loads(response)
            return result
        except:
            # Fallback calculation
            base_score = 650
            if loan_data.get('monthly_income', 0) > 50000:
                base_score += 50
            if loan_data.get('existing_loans', 0) == 0:
                base_score += 40
            if len(user_data.get('past_loans', [])) > 0:
                base_score += 30
            
            risk = "low" if base_score >= 720 else "medium" if base_score >= 650 else "high"
            return {
                "credit_score": min(850, base_score),
                "risk_level": risk,
                "factors": ["Stable income", "Low existing debt", "Good transaction history"]
            }

class DocumentVerificationAgent:
    """Verifies user documents and salary slips"""
    
    async def verify_documents(self, aadhar: str, pan: str, face_image: Optional[str]) -> Dict[str, Any]:
        # Simulate document verification
        return {
            "aadhar_verified": True,
            "pan_verified": True,
            "face_verified": True if face_image else False,
            "verification_score": 95,
            "notes": "All documents verified successfully"
        }
    
    async def analyze_salary_slip(self, monthly_income: float, employment_type: str) -> Dict[str, Any]:
        # Simulate OCR and analysis
        return {
            "verified": True,
            "extracted_income": monthly_income,
            "consistency_score": 92,
            "employment_stability": "high" if employment_type == "salaried" else "medium"
        }

class EMICalculationAgent:
    """Calculates EMI, interest rates, and repayment schedules"""
    
    def calculate_emi(self, principal: float, tenure_months: int, risk_level: str) -> Dict[str, Any]:
        # Interest rate based on risk
        interest_rates = {
            "low": 8.5,
            "medium": 11.0,
            "high": 14.5
        }
        annual_rate = interest_rates.get(risk_level, 11.0)
        monthly_rate = annual_rate / 12 / 100
        
        # EMI Formula: P * r * (1+r)^n / ((1+r)^n - 1)
        if monthly_rate > 0:
            emi = principal * monthly_rate * ((1 + monthly_rate) ** tenure_months) / (((1 + monthly_rate) ** tenure_months) - 1)
        else:
            emi = principal / tenure_months
        
        total_amount = emi * tenure_months
        total_interest = total_amount - principal
        
        return {
            "emi_amount": round(emi, 2),
            "interest_rate": annual_rate,
            "total_amount": round(total_amount, 2),
            "total_interest": round(total_interest, 2),
            "processing_fee": round(principal * 0.01, 2)  # 1% processing fee
        }

class UnderwritingAgent:
    """Makes final loan approval/rejection decision"""
    
    def __init__(self):
        self.chat = LlmChat(
            api_key=MISTRAL_API_KEY,
            session_id=f"underwriting_{datetime.utcnow().timestamp()}",
            system_message="You are a senior loan underwriting officer. Review all analysis and make final loan approval decisions. Return only JSON with keys: decision (approved/rejected), approved_amount, reason, conditions (list)."
        ).with_model("openai", "gpt-5.2")
    
    async def make_decision(self, loan_data: Dict[str, Any], credit_analysis: Dict[str, Any], 
                           doc_verification: Dict[str, Any], emi_calculation: Dict[str, Any]) -> Dict[str, Any]:
        prompt = f"""
Loan Underwriting Review:

Loan Request:
- Amount: ₹{loan_data.get('amount', 0):,.2f}
- Purpose: {loan_data.get('purpose')}
- Tenure: {loan_data.get('tenure_months')} months

Credit Analysis:
- Credit Score: {credit_analysis.get('credit_score')}
- Risk Level: {credit_analysis.get('risk_level')}
- Factors: {', '.join(credit_analysis.get('factors', []))}

Document Verification:
- Verification Score: {doc_verification.get('verification_score')}%
- Status: {'Verified' if doc_verification.get('aadhar_verified') else 'Pending'}

EMI Calculation:
- EMI Amount: ₹{emi_calculation.get('emi_amount', 0):,.2f}
- Interest Rate: {emi_calculation.get('interest_rate')}%
- Total Repayment: ₹{emi_calculation.get('total_amount', 0):,.2f}

Make underwriting decision. Return JSON with: decision, approved_amount, reason, conditions.
"""
        try:
            response = await self.chat.send_message(UserMessage(text=prompt))
            result = json.loads(response)
            return result
        except:
            # Fallback decision (approve all for demo)
            credit_score = credit_analysis.get('credit_score', 650)
            requested_amount = loan_data.get('amount', 0)
            
            if credit_score >= 700:
                return {
                    "decision": "approved",
                    "approved_amount": requested_amount,
                    "reason": "Strong credit profile and stable income",
                    "conditions": ["Maintain account balance", "Timely EMI payments"]
                }
            elif credit_score >= 650:
                return {
                    "decision": "approved",
                    "approved_amount": requested_amount * 0.85,
                    "reason": "Moderate risk profile, approved with reduced amount",
                    "conditions": ["Provide guarantor", "Monthly income proof", "Timely EMI payments"]
                }
            else:
                return {
                    "decision": "approved",  # Demo: approve all
                    "approved_amount": requested_amount * 0.7,
                    "reason": "Higher risk but manageable with conditions",
                    "conditions": ["Provide guarantor", "Life insurance coverage", "Quarterly income verification"]
                }

class LoanOrchestratorAgent:
    """Main orchestrator that coordinates all agents"""
    
    def __init__(self):
        self.credit_agent = CreditScoringAgent()
        self.doc_agent = DocumentVerificationAgent()
        self.emi_agent = EMICalculationAgent()
        self.underwriting_agent = UnderwritingAgent()
    
    async def process_loan_application(self, user_data: Dict[str, Any], loan_data: Dict[str, Any]) -> Dict[str, Any]:
        """Process complete loan application through all agents"""
        
        # Step 1: Credit Scoring
        credit_analysis = await self.credit_agent.calculate_score(user_data, loan_data)
        
        # Step 2: Document Verification
        doc_verification = await self.doc_agent.verify_documents(
            user_data.get('aadhar'),
            user_data.get('pan'),
            user_data.get('face_image')
        )
        
        # Step 3: Salary Slip Analysis
        salary_analysis = await self.doc_agent.analyze_salary_slip(
            loan_data.get('monthly_income'),
            loan_data.get('employment_type')
        )
        
        # Step 4: EMI Calculation
        emi_calculation = self.emi_agent.calculate_emi(
            loan_data.get('amount'),
            loan_data.get('tenure_months'),
            credit_analysis.get('risk_level', 'medium')
        )
        
        # Step 5: Underwriting Decision
        underwriting_decision = await self.underwriting_agent.make_decision(
            loan_data,
            credit_analysis,
            doc_verification,
            emi_calculation
        )
        
        return {
            "credit_analysis": credit_analysis,
            "document_verification": doc_verification,
            "salary_analysis": salary_analysis,
            "emi_calculation": emi_calculation,
            "underwriting_decision": underwriting_decision,
            "processed_at": datetime.utcnow().isoformat()
        }

class ChatbotAgent:
    """AI Chatbot that acts as virtual loan officer"""
    
    def __init__(self, user_id: str):
        self.user_id = user_id
        self.chat = LlmChat(
            api_key=MISTRAL_API_KEY,
            session_id=f"chatbot_{user_id}",
            system_message="""You are a professional virtual loan officer at NBFC Bank. Your role:
- Help users understand loan products and eligibility
- Guide them through application process
- Explain credit scores, interest rates, and EMI calculations
- Negotiate loan terms (amount, tenure, EMI) within reasonable limits
- Provide personalized financial advice
- Be professional, empathetic, and helpful
- Access user's financial data to give accurate information
- You can approve, modify, or suggest alternatives for loan applications

Always be clear, concise, and customer-focused."""
        ).with_model("openai", "gpt-5.2")
    
    async def chat_with_user(self, message: str, user_context: Dict[str, Any]) -> str:
        """Handle user chat with context"""
        context = f"""
User Context:
- Name: {user_context.get('name')}
- Account Balance: ₹{user_context.get('account_balance', 0):,.2f}
- Credit Score: {user_context.get('credit_score', 'Not calculated yet')}
- Active Loans: {user_context.get('active_loans', 0)}
- Past Loans: {user_context.get('past_loans_count', 0)}

User Message: {message}
"""
        response = await self.chat.send_message(UserMessage(text=context))
        return response